Add tests for critical section HAL API

maciejbocianski opened the pull request:
## Description

This PR contains critical section HAL API tests.
Test were created based on documentation provided in PR #5346 

## Status

**READY**

## Migrations

NO

## Notes
~~in order to run these tests #5346 has to be merged first,~~

## Todos

- [x] adopt test for NRF5X series

maciejbocianski commented:
@scartmell-arm 
as discussed it would be good to have `hal_in_critical_section` function to cover NRF5x like targets which mask application interrupts rather than disable all interrupts

maciejbocianski commented:
support for NRF5x added

0xc0170 commented on TESTS/mbed_hal/critical_section/main.cpp:
```diff
@@ -0,0 +1,84 @@
+{
+    // NRF5x targets don't disable interrupts when in critical section, instead they mask application interrupts this is due to BLE stack
+    // (BLE to be operational requires some interrupts to be always enabled)
+#ifdef TARGET_NRF52_DK
```
wasn't there somewhere a request for having function to check if interrupts are enabled? Or even an implementation?

maciejbocianski replied:
but it would be good toindeed, there is function `hal_in_critical_section()`, but it would be nice to verify API functions by something form outside critical section API. `us_ticker` could be used to check interrupts but unfortunately it utilises `core_util_critical_section_` (which also call HAL critical API ) and it brakes the test. Generly we should always use platform `core_util_critical_section` not HAL, in particular we cannot use both in the same time

0xc0170 replied:
I was concerned about the portability (how this function can grow in time).

Looks fine to test this here.

0xc0170 commented:
/morph build

cmonr commented:
/morph build

cmonr commented:
Jenkins didn't report abort.

/morph export-build

maciejbocianski commented:
@scartmell-arm @0xc0170 
I have added fix for masked interrupts check on NRF52

It turned out that none of interrupts handled by `ISER/ICER[1]` is actually enabled(used)
and `(NVIC->ISER[1] & __NRF_NVIC_APP_IRQS_1) != 0` always returns `false`.


I modified the checking code from:
```
return (((NVIC->ISER[0] & __NRF_NVIC_APP_IRQS_0) != 0) 
           && ((NVIC->ISER[1] & __NRF_NVIC_APP_IRQS_1) != 0));
```
to:
```
return (((NVIC->ISER[0] & __NRF_NVIC_APP_IRQS_0) != 0) 
           || ((NVIC->ISER[1] & __NRF_NVIC_APP_IRQS_1) != 0));
```

0xc0170 commented:
/morph build

maciejbocianski commented:
/morph test

0xc0170 commented:
/morph build
